src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java

```diff
@@ -220,29 +221,100 @@ public void doNothing(String mnemonic, Object inputs) throws EvalException {
   static final GeneratedMessage.GeneratedExtension<ExtraActionInfo, SpawnInfo> SPAWN_INFO =
       SpawnInfo.spawnInfo;
 
+  private String progressMessageForSymlink(
+      Object /* String or None */ progressMessage, Artifact output) {
+    if (progressMessage != Starlark.NONE) {
+      // Should have been verified by Starlark before this function is called.
+      return (String)progressMessage;
+    }
+
+    return "Creating symlink " + output.getRootRelativePathString();
+  }
+
+  /** Return whether exactly one of the passed object is not {@link Starlark.NONE}. */
+  private boolean exactlyOneNotNone(Object first, Object... others) {
+    boolean seen = first != Starlark.NONE;
+    for (Object o : others) {
+      if (o == Starlark.NONE) {
+        continue;
+      }
+      if (seen) {
+        return false;
+      }
+      seen = true;
+    }
+    return seen;
+  }
+
   @Override
-  public void symlink(FileApi output, String path) throws EvalException {
+  public void symlink(
+      FileApi output,
+      Object /* Artifact or None */ targetFile,
+      Object /* String or None */ targetPath,
+      Boolean isExecutable,
+      Object /* String or None */ progressMessage)
+      throws EvalException {
     context.checkMutable("actions.symlink");
 
+    if (!exactlyOneNotNone(targetFile, targetPath)) {
+      throw Starlark.errorf("\"target_file\" and \"target_path\" cannot be set at the same time");
+    }
+
+    // Should have been verified by Starlark before this function is called.
+    Artifact outputArtifact = (Artifact)output;
+
+    if (targetFile != Starlark.NONE) {
+      if (outputArtifact.isSymlink()) {
+        throw Starlark.errorf(
+            "output of symlink action with \"target_file\" must be created by declare_file()");
+      }
+
+      // Should have been verified by Starlark before this function is called.
+      Artifact target = (Artifact)targetFile;
+      if (isExecutable) {
+        SymlinkAction action =
+            SymlinkAction.toExecutable(
+                ruleContext.getActionOwner(),
+                target,
+                outputArtifact,
+                progressMessageForSymlink(progressMessage, outputArtifact));
+        registerAction(action);
+        return;
+      }
+      SymlinkAction action =
+          SymlinkAction.toArtifact(
+              ruleContext.getActionOwner(),
+              target,
```

Comment:
Question: If `symlink()` makes an `Artifact` with `isSymlink() == false` point to another `Artifact` with `isSymlink() == false`, doesn't that mean you can have two symlinks point to each other? Or is that prevented by having the target `Artifact` considered an input to the symlink action?